Mathlib/MeasureTheory/Integral/Layercake.lean

```diff
@@ -377,3 +381,95 @@ theorem lintegral_rpow_eq_lintegral_meas_lt_mul (μ : Measure α) [SigmaFinite 
+  have markov := mul_meas_ge_le_lintegral₀ (μ := μ) norm_f_aemble (ENNReal.ofReal c)
+  have obs : ∫⁻ (a : α), ENNReal.ofReal ‖f a‖ ∂μ = ∫⁻ (a : α), ‖f a‖₊ ∂μ :=
+    lintegral_congr (fun x ↦ ofReal_norm_eq_coe_nnnorm (f x))
+  simp_rw [ENNReal.ofReal_le_ofReal_iff (norm_nonneg _), obs] at markov
+  have c_inv_ne_top : (ENNReal.ofReal c)⁻¹ ≠ ∞ :=
+    ENNReal.inv_ne_top.mpr (ENNReal.ofReal_pos.mpr c_pos).ne.symm
+  simpa [← mul_assoc,
+         ENNReal.inv_mul_cancel (ENNReal.ofReal_pos.mpr c_pos).ne.symm ENNReal.ofReal_ne_top]
+    using ENNReal.mul_lt_top c_inv_ne_top (lt_of_le_of_lt markov f_intble.2).ne
+
+/-- If `f` is integrable, then for any `c > 0` the set `{x | ‖f x‖ > c}` has finite measure. -/
+lemma Integrable.measure_const_lt_norm_lt_top
+    {E : Type*} [NormedAddCommGroup E] [MeasurableSpace E] [BorelSpace E]
+    {f : α → E} (f_intble : Integrable f μ) {c : ℝ} (c_pos : 0 < c) :
+    μ {a : α | c < ‖f a‖} < ∞ :=
+  lt_of_le_of_lt (measure_mono (fun _ h ↦ (Set.mem_setOf_eq ▸ h).le))
+    (Integrable.measure_const_le_norm_lt_top f_intble c_pos)
+
+variable {f : α → ℝ} (f_intble : Integrable f μ)
+
+/-- If `f` is `ℝ`-valued and integrable, then for any `c > 0` the set `{x | f x ≥ c}` has finite
+measure. -/
+lemma Integrable.measure_const_le_lt_top {c : ℝ} (c_pos : 0 < c) :
+    μ {a : α | c ≤ f a} < ∞ := by
+  refine lt_of_le_of_lt (measure_mono ?_) (Integrable.measure_const_le_norm_lt_top f_intble c_pos)
+  intro x hx
+  simp only [Real.norm_eq_abs, Set.mem_setOf_eq] at hx ⊢
+  exact hx.trans (le_abs_self _)
+
+/-- If `f` is `ℝ`-valued and integrable, then for any `c < 0` the set `{x | f x ≤ c}` has finite
+measure. -/
+lemma Integrable.measure_le_const_lt_top {c : ℝ} (c_neg : c < 0) :
+    μ {a : α | f a ≤ c} < ∞ := by
+  refine lt_of_le_of_lt (measure_mono ?_)
+          (Integrable.measure_const_le_norm_lt_top f_intble (show 0 < -c by linarith))
+  intro x hx
+  simp only [Real.norm_eq_abs, Set.mem_setOf_eq] at hx ⊢
+  exact (show -c ≤ - f x by linarith).trans (neg_le_abs_self _)
+
+/-- If `f` is `ℝ`-valued and integrable, then for any `t > 0` the set `{x | f x > t}` has finite
+measure. -/
+lemma Integrable.measure_const_lt_lt_top {c : ℝ} (c_pos : 0 < c) :
+    μ {a : α | c < f a} < ∞ :=
+  lt_of_le_of_lt (measure_mono (fun _ hx ↦ (Set.mem_setOf_eq ▸ hx).le))
+    (Integrable.measure_const_le_lt_top f_intble c_pos)
+
+/-- If `f` is `ℝ`-valued and integrable, then for any `t < 0` the set `{x | f x < t}` has finite
+measure. -/
+lemma Integrable.measure_lt_const_lt_top {c : ℝ} (c_neg : c < 0) :
+    μ {a : α | f a < c} < ∞ :=
+  lt_of_le_of_lt (measure_mono (fun _ hx ↦ (Set.mem_setOf_eq ▸ hx).le))
+    (Integrable.measure_le_const_lt_top f_intble c_neg)
+
+/-- The standard case of the layer cake formula / Cavalieri's principle / tail probability formula:
+
+For an integrable a.e.-nonnegative real-valued function `f` on a sigma-finite measure space,
```

Comment:
again, I don't see why sigma-finite should be relevant here.

Reply:
In this actual layercake formula result, sigma-finiteness is needed because the proof of the layercake formula is at its heart an application of Fubini's theorem.

(One can trace back the necessity of the hypotheses in the current proof to first the most fundamental layercake formula `lintegral_comp_eq_lintegral_meas_le_mul_of_measurable` in the present file, and in its proof to Fubini's theorems `lintegral_lintegral_swap`).

You are completely right that including `[SigmaFinite μ]` in the `variable` line in this section is a mistake, none of the results except this very last one need this hypothesis.

Reply:
Still I don't think that sigma-finiteness is relevant here: you could restrict your measure to the set where `f > 0` (which doesn't change anything to the desired formula), and this restricted measure is sigma-finite as the function is integrable. Maybe first state an auxiliary lemma assuming the measure is sigma-finite, and deduce the general case afterwards?
(See `MeasureTheory.Integrable.aefinStronglyMeasurable` and `AEFinStronglyMeasurable.exists_set_sigmaFinite`)

Reply:
Oh, thanks, this seems like a nice strenghtening!

Reply:
I wrote a proof without the sigma-finiteness assumption as you suggested.

The proof structure is in principle the same as before, except two observations `obs` and `obs'` are needed to reduce to the earlier case. Mathematically these two observations are trivial, but I failed to come up with short proofs. I'm probably missing some good ways of handling the a.e. properties and restrictions of measures here.

Reply:
Can you remove `sigma-finite` from the docstring then?

Reply:
Sorry, I forgot to update the docstring! I now fixed the docstring of this theorem and added a note in the module docstring. The module docstring starts by assuming sigma-finiteness, and I'd keep it that way unless you think it is misleading.

Thanks again! Getting rid of the sigma-finiteness assumption is a nice generalization that simply did not occur to me when writing this.

And sorry, still, that my proofs of `obs` and `obs'` are clumsy. Both sound like they should be one-liners and then they should be done in the middle of the proof, but if they are not one-liners, should they be separated to lemmas of their own? At least `obs` might have other uses. For `obs'`, I'd have to think about a formulation that has potential of being useful elsewhere. I will still at least think about this, but if you have a clear opinion about whether to separate these, I'll be happy to hear.

Reply:
I will rewrite separating these as lemmas --- I believe it will be clearer and the golfing opportunities will also be clearer. I marked as `awaiting_author` while I do the change.